src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -370,61 +449,72 @@ public TStructuralType CopyChangedValues(TStructuralType original)
 
             CopyChangedDynamicValues(original);
 
-            // For nested resources.
-            foreach (string nestedResourceName in _deltaNestedResources.Keys)
-            {
-                // Patch for each nested resource changed under this TStructuralType.
-                dynamic deltaNestedResource = _deltaNestedResources[nestedResourceName];
-                dynamic originalNestedResource = null;
-                if (!TryGetPropertyRef(original, nestedResourceName, out originalNestedResource))
-                {
-                    throw Error.Argument(nestedResourceName, SRResources.DeltaNestedResourceNameNotFound,
-                        nestedResourceName, original.GetType());
-                }
-
-                if (originalNestedResource == null)
-                {
-                    // When patching original target of null value, directly set nested resource.
-                    dynamic deltaObject = _deltaNestedResources[nestedResourceName];
-                    dynamic instance = deltaObject.GetInstance();
+            CopyChangedNestedProperties(original);
+        }
 
-                    // Recursively patch up the instance with the nested resources.
-                    deltaObject.CopyChangedValues(instance);
+        /// <summary>
+        /// Copies the unchanged property values from the underlying entity (accessible via <see cref="GetInstance()" />)
+        /// to the <paramref name="original"/> entity.
+        /// </summary>
+        /// <param name="original">The entity to be updated.</param>
+        public void CopyUnchangedValues(TStructuralType original)
+        {
+            if (original == null)
+            {
+                throw Error.ArgumentNull(nameof(original));
+            }
 
-                    _allProperties[nestedResourceName].SetValue(original, instance);
-                }
-                else
-                {
-                    // Recursively patch the subtree.                    
-                    Contract.Assert(TypedDelta.IsDeltaOfT(((object)deltaNestedResource).GetType()), nestedResourceName + "'s corresponding value should be Delta<T> type but is not.");
+            if (!_structuredType.IsInstanceOfType(original))
+            {
+                throw Error.Argument(nameof(original), SRResources.DeltaTypeMismatch, _structuredType, original.GetType());
+            }
 
-                    Type newType = deltaNestedResource.StructuredType;
-                    Type originalType = originalNestedResource.GetType();
+            IEnumerable<PropertyAccessor<TStructuralType>> propertiesToCopy = GetUnchangedPropertyNames().Select(s => _allProperties[s]);
```

Comment:
Should we be intersecting `_allProperties` with updateable properties here?

Reply:
Yes we should and we are intersecting. The `GetUnchangedPropertyNames` method already does the intersection.